# Latihan Praktikum Damin Modul 5
Nama: Elok Fiola    
NIM:  122450051        
Kelas: RC

# code untuk metode seleksi maju

In [ ]:
#Impor Library
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.feature_selection import SequentialFeatureSelector
from sklearn.metrics import accuracy_score
penguins = pd.read_csv('penguins.csv')

# Drop rows with missing values
penguins = penguins.dropna()

#Menentukan Fitur dan Target:
X = penguins[['bill_length', 'bill_depth', 'flipper_length', 'body_mass']]
y = penguins['species']

#Membagi Data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Model yang digunakan
model = RandomForestClassifier(random_state=42)

# Seleksi Maju (Forward Selection)
sfs = SequentialFeatureSelector(model, n_features_to_select=2, direction='forward')
sfs.fit(X_train, y_train)
print("Fitur yang dipilih:", sfs.get_support(indices=True))

#Transformasi Data
X_train_selected = sfs.transform(X_train)
X_test_selected = sfs.transform(X_test)

#Pelatihan Model dan Prediksi
model.fit(X_train_selected, y_train)
y_pred = model.predict(X_test_selected)

#Menghitung Akurasi
accuracy = accuracy_score(y_test, y_pred)
print("Akurasi model dengan fitur yang dipilih:", accuracy)

Fitur yang dipilih: [0 2]
Akurasi model dengan fitur yang dipilih: 0.9850746268656716


Ini menunjukkan bahwa fitur yang terpilih adalah fitur dengan indeks 0 (panjang paruh) dan indeks 2 (panjang sirip). Ini berarti model berfokus pada dua atribut tersebut untuk melakukan klasifikasi spesies penguin.     

model memiliki akurasi sekitar 98.51% pada data pengujian, yang berarti model sangat baik dalam mengklasifikasikan spesies penguin berdasarkan fitur yang dipilih.


# code untuk metode seleksi mundur

In [ ]:
from sklearn.feature_selection import SequentialFeatureSelector

# Seleksi Mundur (Backward Selection)
sbs = SequentialFeatureSelector(model, n_features_to_select=2, direction='backward')
sbs.fit(X, y)

# Fitur yang dipilih
print("Fitur yang dipilih:", sbs.get_support(indices=True))
X_selected = sbs.transform(X) #Transformasi Data
model.fit(X_selected, y) #Pelatihan Model

Fitur yang dipilih: [0 1]


RandomForestClassifier(random_state=42)

menunjukkan bahwa fitur yang terpilih adalah fitur dengan indeks 0 (panjang paruh) dan indeks 1 (kedalaman paruh). Ini berarti model memilih dua fitur ini sebagai yang paling relevan untuk klasifikasi spesies penguin.

# code untuk metode seleksi dua arah

In [ ]:
sfs_bidirectional = SequentialFeatureSelector(model, n_features_to_select=2,

direction='forward', tol=1e-4)

sfs_bidirectional.fit(X, y)
# Fitur yang dipilih
print("Fitur yang dipilih:", sfs_bidirectional.get_support(indices=True))
# Evaluasi model dengan fitur yang dipilih
X_selected = sfs_bidirectional.transform(X)
model.fit(X_selected, y)

Fitur yang dipilih: [0 2]


RandomForestClassifier(random_state=42)

# Code Untuk Metode Random Generation

In [ ]:
import numpy as np
from itertools import combinations
from sklearn.metrics import accuracy_score

# Pilih fitur secara acak
n_features = X.shape[1]
all_combinations = list(combinations(range(n_features), 2)) # Kombinasi 2 fitur
best_accuracy = 0
best_combination = None

for comb in all_combinations:
  # Use .iloc to select columns by integer index
  X_selected = X.iloc[:, list(comb)]
  model.fit(X_selected, y)
  y_pred = model.predict(X_selected)
  acc = accuracy_score(y, y_pred)
  if acc > best_accuracy:
    best_accuracy = acc
    best_combination = comb
print(f"Kombinasi terbaik: {best_combination} dengan akurasi {best_accuracy}")

Kombinasi terbaik: (0, 1) dengan akurasi 1.0


# Latihan

Menggunakan maju, mundur, dan dua arah,

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.feature_selection import SequentialFeatureSelector
from itertools import combinations
# Load dataset segmentation-all.csv ( pastikan file tersebut ada dalam direktori yang sama )
data = pd.read_csv('segmentation-all.csv')
# Pisahkan fitur dan target
X = data.drop('Class', axis=1)
y = data['Class']
# Bagi data menjadi data latih dan data uji
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
# Model yang digunakan
model = RandomForestClassifier(random_state=42)

# Seleksi Maju (Forward Selection)
sfs = SequentialFeatureSelector(model, n_features_to_select=5, direction='forward')
sfs.fit(X_train, y_train)
# Fitur yang dipilih

print("Fitur yang dipilih (Forward Selection):", sfs.get_support(indices=True))
# Evaluasi model dengan fitur yang dipilih
X_train_selected = sfs.transform(X_train)
X_test_selected = sfs.transform(X_test)
model.fit(X_train_selected, y_train)
y_pred = model.predict(X_test_selected)
accuracy = accuracy_score(y_test, y_pred)
print("Akurasi model dengan fitur yang dipilih (Forward Selection):", accuracy)

# Seleksi Mundur (Backward Selection)
sbs = SequentialFeatureSelector(model, n_features_to_select=5, direction='backward')
sbs.fit(X_train, y_train)
# Fitur yang dipilih
print("\nFitur yang dipilih (Backward Selection):", sbs.get_support(indices=True))
# Evaluasi model dengan fitur yang dipilih
X_train_selected = sbs.transform(X_train)
X_test_selected = sbs.transform(X_test)
model.fit(X_train_selected, y_train)
y_pred = model.predict(X_test_selected)
accuracy = accuracy_score(y_test, y_pred)
print("Akurasi model dengan fitur yang dipilih (Backward Selection):", accuracy)

# Seleksi dua arah (Bidirectional Selection)
sfs_bidirectional = SequentialFeatureSelector(model, n_features_to_select=5, direction='forward')
sfs_bidirectional.fit(X_train, y_train)
# Fitur yang dipilih
print("\nFitur yang dipilih (Bidirectional Selection):", sfs_bidirectional.get_support(indices=True))
# Evaluasi model dengan fitur yang dipilih
X_train_selected = sfs_bidirectional.transform(X_train)
X_test_selected = sfs_bidirectional.transform(X_test)
model.fit(X_train_selected, y_train)
y_pred = model.predict(X_test_selected)
accuracy = accuracy_score(y_test, y_pred)
print("Akurasi model dengan fitur yang dipilih (Bidirectional Selection):", accuracy)

# Random Generation
n_features = X.shape[1]
all_combinations = list(combinations(range(n_features), 5)) # Kombinasi 5 fitur
best_accuracy = 0
best_combination = None
for comb in all_combinations:
  # Use .iloc to select columns by integer index
  X_selected = X.iloc[:, list(comb)]
  X_train_selected, X_test_selected, y_train, y_test = train_test_split(X_selected, y, test_size=0.2)
  model.fit(X_train_selected, y_train)
  y_pred = model.predict(X_test_selected)
  acc = accuracy_score(y_test, y_pred)
  if acc > best_accuracy:
    best_accuracy = acc
    best_combination = comb
print("\nKombinasi terbaik (Random Generation):", best_combination, "dengan akurasi", best_accuracy)

Fitur yang dipilih (Forward Selection): [ 0  1  6 10 18]
Akurasi model dengan fitur yang dipilih (Forward Selection): 0.987012987012987

Fitur yang dipilih (Backward Selection): [ 0  1 10 17 18]
Akurasi model dengan fitur yang dipilih (Backward Selection): 0.9848484848484849

Fitur yang dipilih (Bidirectional Selection): [ 0  1  6 10 18]
Akurasi model dengan fitur yang dipilih (Bidirectional Selection): 0.987012987012987


KeyboardInterrupt: 

model mencapai akurasi sekitar 98.70% pada data uji. Akurasi yang tinggi ini menunjukkan bahwa fitur yang dipilih sangat informatif dan model cukup baik dalam memprediksi kelas dari data yang diberikan.     


